File: examples/python/query.py
"""Run a single SQL statement against the Quack-on-Demand FlightSQL edge and
print the Arrow result.

Run:  python query.py
      python query.py "SELECT count(*) FROM tpch1.customer"

See client.py for the ADBC plumbing. Connection settings come from QOD_* env
vars (see README).
"""

import json
import os
import sys

from client import QodClient, config_from_env, row_to_dict

SQL = (
    sys.argv[1]
    if len(sys.argv) > 1
    else os.environ.get(
        "QOD_SQL",
        "SELECT * FROM (VALUES (1, 'duck'), (2, 'quack'), (3, 'lake')) AS t(id, label)",
    )
)


def main() -> None:
    with QodClient.connect(config_from_env()) as client:
        print(f"Connecting to {client.describe()}")
        print(f"SQL: {SQL}\n")

        table = client.query(SQL)

        columns = ", ".join(f"{f.name}: {f.type}" for f in table.schema)
        print(f"columns: {columns}")
        print(f"rows: {table.num_rows}\n")
        for i in range(min(table.num_rows, 100)):
            print(json.dumps(row_to_dict(table, i)))


if __name__ == "__main__":
    try:
        main()
    except Exception as err:  # noqa: BLE001 - top-level CLI error reporting
        print(f"\nquery failed: {err}", file=sys.stderr)
        sys.exit(1)

File: examples/typescript/src/query.ts
// Run a single SQL statement against the Quack-on-Demand FlightSQL edge and
// print the Arrow result.
//
// Run:  npm run query
//       npm run query -- "SELECT count(*) FROM tpch1.customer"
//
// See client.ts for the gRPC + Arrow plumbing and src/proto/ for the protocol
// subset. Connection settings come from QOD_* env vars (see README).
import { QodClient, configFromEnv, rowToObject } from "./client.js";

const SQL =
  process.argv[2] ??
  process.env.QOD_SQL ??
  "SELECT * FROM (VALUES (1, 'duck'), (2, 'quack'), (3, 'lake')) AS t(id, label)";

async function main(): Promise<void> {
  const client = await QodClient.connect(configFromEnv());
  console.log(`Connecting to ${client.describe()}`);
  console.log(`SQL: ${SQL}\n`);

  const table = await client.query(SQL);

  const columns = table.schema.fields.map((f) => `${f.name}: ${f.type}`).join(", ");
  console.log(`columns: ${columns}`);
  console.log(`rows: ${table.numRows}\n`);
  for (const row of table.toArray().slice(0, 100)) {
    console.log(JSON.stringify(rowToObject(row, table)));
  }

  client.close();
}

main().catch((err) => {
  console.error("\nquery failed:", err?.message ?? err);
  process.exit(1);
});